### PR TITLE
Documentation: Add use case examples for the new `object_keys` function

### PR DESCRIPTION
**Pull request by amotl**
This adds two exercises for the new `object_keys` function.
- ad88d2d6: Add example how to match record columns for empty or unassigned objects. References:
  - #12714 
  - #9398
- 2183302de69: Outline general usage within the storyline-like `dql/selects.rst`. One example displays the output, the other one uses it as a filter expression, combined with the `ANY` operator.

### CODE REVIEW

File: docs/general/dql/selects.rst [hunk 1]
```diff
@@ -834,6 +834,42 @@ Object property can also be addressed in the :ref:`where clause
...
+.. Hidden: refresh locations
+
+    cr> refresh table locations;
+    REFRESH OK, 1 row affected (... sec)
+
+a suitable expression to match for both ``NULL`` values and empty objects, is::
+
+    cr> select count(*) from locations where
+    ... inhabitants is null or
+    ... inhabitants = {};
```
Comment (seut) [inline docs/general/dql/selects.rst:852]: `object_col = {}` does not match empty objects. But `IS NULL` does. So I'd think we don't need to document this for objects as  `IS NULL` behaves like expected.

File: docs/general/dql/selects.rst [hunk 2]
```diff
@@ -834,6 +834,42 @@ Object property can also be addressed in the :ref:`where clause
...
+
+    cr> select count(*) from locations where
+    ... array_length(object_keys(inhabitants), 1) = 2;
+    +----------+
+    | count(*) |
+    +----------+
+    |        1 |
+    +----------+
+    SELECT 1 row in set (... sec)
+
```
Comment (seut) [inline docs/general/dql/selects.rst:872]: Shall we maybe drop the new rows in a hidden statement so later result sets must not me adjusted?

Comment (amotl) [reply docs/general/dql/selects.rst:877]: I've amended the patch to do it like you suggested.

**Review by seut — APPROVED** [review]

File: docs/general/dql/selects.rst (hunk 1 above)
Comment (amotl) [reply docs/general/dql/selects.rst]: I've changed the statement to only use the `IS NULL` match.

File: docs/general/dql/selects.rst [hunk 3]
```diff
@@ -841,6 +841,50 @@ Object property can also be addressed in the :ref:`where clause
...
+<scalar-array_length>` scalar functions::
+
+    cr> select count(*) from locations where
+    ... array_length(object_keys(inhabitants), 1) = 2;
+    +----------+
+    | count(*) |
+    +----------+
+    |        1 |
+    +----------+
+    SELECT 1 row in set (... sec)
```
Comment (amotl) [inline docs/general/dql/selects.rst:881]: This section is now both nonsense and confusing. I will replace it with a better example to demonstrate the usage of the new `object_keys()` function.

File: docs/general/dql/selects.rst [hunk 4]
```diff
@@ -841,6 +841,50 @@ Object property can also be addressed in the :ref:`where clause
...
+
+.. Hidden: drop record again
+
+    cr> delete from locations where id=99;
+    DELETE OK, 1 row affected  (... sec)
+
+For determining the number of items in an object, you can use the
+:ref:`object_keys <scalar-object_keys>` and :ref:`array_length
+<scalar-array_length>` scalar functions::
```
Comment (mfussenegger) [inline docs/general/dql/selects.rst:872]: I think there should be some elaboration why `IS NULL` matches on empty objects. I'm also not sure if this is generally the case (e.g. sys tables, or if the column doesn't have an index)

**Review by mfussenegger — CHANGES_REQUESTED** [review]

File: docs/general/dql/selects.rst [hunk 5]
```diff
@@ -841,6 +841,50 @@ Object property can also be addressed in the :ref:`where clause
...
+    +----------+
+    |        1 |
+    +----------+
+    SELECT 1 row in set (... sec)
+
+.. NOTE::
+
+    ``array_length(object_keys(inhabitants), 1)`` will return ``NULL`` on empty
+    objects.
+
```
Comment (mfussenegger) [inline docs/general/dql/selects.rst:887]:
```suggestion
```

Seems unrelated and doesn't explain why this is the case.

Comment (seut) [reply docs/general/dql/selects.rst:887]: Maybe not good explained, but it worth mentioning somewhere (maybe better at the `array_length` documentation), that `array_length(some_array, 1)` will return `NULL` on empty arrays while someone may expect `0`.

Comment (amotl) [reply docs/general/dql/selects.rst:887]: @mfussenegger: Yes, I will rework this patch and remove unrelated and confusing examples. @seut: I agree and will amend the other patch #12688 correspondingly.

File: docs/general/dql/selects.rst (hunk 4 above)
Comment (amotl) [reply docs/general/dql/selects.rst:872]: Judging from your recent adjustment #12704, I am assuming the same ambiguity might get fixed on this end as well. So I will adjust the patch to use a different expression instead.
```sql
WHERE inhabitants IS NULL OR inhabitants = {}
```

File: docs/general/dql/selects.rst [hunk 6]
```diff
@@ -797,6 +825,58 @@ Object property can also be addressed in the :ref:`where clause
...
+    +------------+-------------------------+
+    | Betelgeuse | ["name", "description"] |
+    +------------+-------------------------+
+    SELECT 1 row in set (... sec)
+
+By combining the :ref:`scalar-object_keys` function with the ``ANY`` operator,
+you can find all records where an object contains a specific object key::
+
+    cr> select name from locations
+    ... where 'interests' = ANY(object_keys(inhabitants));
```
Comment (mfussenegger) [inline docs/general/dql/selects.rst:843]: Doing a `inhabitants['interests'] IS NOT NULL` is probably faster, using `ANY(<somescalar>)` cannot utilize the index.

Comment (amotl) [reply docs/general/dql/selects.rst:843]: Thank you. Your suggestions are very reasonable.

> We could consider having some sort of "working with arrays and objects in CrateDB" tutorial or guide.

That's a good idea.
